### PR TITLE
feat(CL): pool id routing

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #3238

## What is the purpose of the change

Finalized the pool id routing in the x/swaprouter module.

See https://github.com/osmosis-labs/osmosis/issues/3238 for details

## Brief Changelog

- implement a new store index and helpers
- set the mapping from pool id to pool type in `CreatePool()`
- test that it is set correctly
- get the mapping in `RouteExactAmountIn()` and `RouteExactAmountOut()` and translate to the desired swap module
- test that balancer pool is retrieved correctly
- add upgrade handler tests

## Testing and Verifying

- added new tests, including upgrade handler
- also covered by e2e

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/swaprouter/creator_test.go [hunk 1]
```diff
@@ -22,20 +24,23 @@ func (suite *KeeperTestSuite) TestCreatePool() {
...
+			expectedModuleType: gammKeeperType,
 		},
 		{
-			name:              "second balancer pool - success",
-			creatorFundAmount: sdk.NewCoins(sdk.NewCoin(denomA, defaultInitPoolAmount.Mul(sdk.NewInt(2))), sdk.NewCoin(denomB, defaultInitPoolAmount.Mul(sdk.NewInt(2)))),
-			msg:               validBalancerPoolMsg,
+			name:               "second balancer pool - success",
+			creatorFundAmount:  sdk.NewCoins(sdk.NewCoin(denomA, defaultInitPoolAmount.Mul(sdk.NewInt(2))), sdk.NewCoin(denomB, defaultInitPoolAmount.Mul(sdk.NewInt(2)))),
+			msg:                validBalancerPoolMsg,
+			expectedModuleType: gammKeeperType,
```
Comment (mattverse) [inline x/swaprouter/creator_test.go:43]: What's the significance of having first and second msg? Do they not get initialized anyways?

File: x/swaprouter/keeper_test.go [hunk 2]
```diff
@@ -31,10 +31,25 @@ func (suite *KeeperTestSuite) SetupTest() {
...
+		// TODO
+		return
+	}
+}
+
+// createBalancerPoolsFromCoins creates balancer pools from given sets of coins.
 // Where element 1 of the input corresponds to the first pool created,
 // element 2 to the second pool created, up until the last element.
-func (suite *KeeperTestSuite) CreateBalancerPoolsFromCoins(poolCoins []sdk.Coins) {
+func (suite *KeeperTestSuite) createBalancerPoolsFromCoins(poolCoins []sdk.Coins) {
```
Comment (mattverse) [inline x/swaprouter/keeper_test.go:52]: Why make testing into private method?

File: x/swaprouter/types/keys.go [hunk 3]
```diff
@@ -11,4 +18,35 @@ const (
...
+	SwapModuleRouterPrefix = []byte{0x02}
 )
+
+// ModuleRouteToBytes serializes moduleRoute to bytes.
+func ModuleRouteToBytes(moduleRoute ModuleRoute) []byte {
+	bz, err := proto.Marshal(&moduleRoute)
+	if err != nil {
+		panic(err)
+	}
+	return bz
```
Comment (mattverse) [inline x/swaprouter/types/keys.go:32]: Is this currently used somewhere? If not, where / how is it planning to get used?

**Review by mattverse — APPROVED** [review]
LGTM! Nice work! 🌮

File: x/swaprouter/creator.go [hunk 4]
```diff
@@ -122,3 +124,11 @@ func (k Keeper) validateCreatedPool(
 	}
 	return nil
 }
+
+// SetModuleRoute stores the mapping from poolId to the given pool type.
+// TODO: unexport after concentrated-liqudity upfrade. Currently, it is exported
```
Comment (czarcas7ic) [inline x/swaprouter/creator.go:129]:
```suggestion
// TODO: unexport after concentrated-liqudity upgrade. Currently, it is exported
```

File: x/swaprouter/router.go [hunk 5]
```diff
@@ -152,9 +156,30 @@ func createMultihopExpectedSwapOuts(
 	return insExpected, nil
 }
 
-func (k Keeper) withSwapModule(isGamm bool) types.SwapI {
-	if isGamm {
-		return k.gammKeeper
+// getSwapModule returns the swap module for the given pool ID.
```
Comment (czarcas7ic) [inline x/swaprouter/router.go:159]:
```suggestion
// GetSwapModule returns the swap module for the given pool ID.
```

File: x/swaprouter/router.go [hunk 6]
```diff
@@ -152,9 +156,30 @@ func createMultihopExpectedSwapOuts(
...
+// Returns error if:
+// - any database error occurs.
+// - fails to find a pool with the given id.
+// - the swap module of the type corresponding to the pool id is not registered
+// in swaprouter's keeper constructor.
+// TODO: unexport after concentrated-liqudity upfrade. Currently, it is exported
```
Comment (czarcas7ic) [inline x/swaprouter/router.go:165]:
```suggestion
// TODO: unexport after concentrated-liqudity upgrade. Currently, it is exported
```

File: x/swaprouter/router_test.go [hunk 7]
```diff
@@ -159,3 +164,68 @@ func (suite *KeeperTestSuite) TestMultihopSwapExactAmountOut() {
...
+		// TODO: valid concentrated liquidity test case.
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		suite.Run(name, func() {
+			suite.SetupTest()
+			swaprouterKeeper := suite.App.SwapRouterKeeper
+
+			suite.createPoolFromType(tc.preCreatePoolType)
```
Comment (czarcas7ic) [inline x/swaprouter/router_test.go:211]: this seems odd, can we just create a single pool prior to running the test cases? from the looks of it, it feels like we should be creating a second balancer pool here with poolId 2 and should not fail (I know thats not how this works, but just is how it looks like it should work on first glance)

**Review by czarcas7ic — APPROVED** [review]
LGTM!

File: x/swaprouter/router_test.go (hunk 7 above)
Comment (p0mvn) [reply x/swaprouter/router_test.go:211]: Why?

We need to test all pool types eventually. We also need to be able to control which kind of pool we are creating from the test case parameter. This helps to achieve that.

File: x/swaprouter/keeper_test.go (hunk 2 above)
Comment (p0mvn) [reply x/swaprouter/keeper_test.go:52]: It is meant to be used only inside swaprouter tests and  cannot be used anywhere else. If we want this globally, we would have to define it in  `apptesting`

File: x/swaprouter/types/keys.go (hunk 3 above)
Comment (p0mvn) [reply x/swaprouter/types/keys.go:32]: Nice catch, this is something I was trying at the start but ended up moving away from. Removed

File: proto/osmosis/swaprouter/v1beta1/module_route.proto [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  Concentrated = 2;
+}
+
+// ModuleRouter defines a route encapsulating pool type.
+// It is used as the value of a mapping from pool id to the pool type,
+// allowing the swap router to know which module to route swaps to given the
+// pool id.
+message ModuleRoute {
+  // pool_type specifies the type of the pool
+  PoolType pool_type = 2;
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply proto/osmosis/swaprouter/v1beta1/module_route.proto:28]: addressed locally

**Comment by p0mvn** [comment]
Addressed all comments

**Pull request merged**